SDL/Kernels.cu

```diff
@@ -68,16 +68,16 @@ __global__ void createMiniDoubletsInGPU(struct SDL::modules& modulesInGPU, struc
 
             if(success)
             {
-                unsigned int mdModuleIndex = atomicAdd(&mdsInGPU.nMDs[lowerModuleIndex],1);
-                if(mdModuleIndex >= N_MAX_MD_PER_MODULES)
+                atomicAdd(&mdsInGPU.totOccupancyMDs[lowerModuleIndex],1);
+                if(mdsInGPU.nMDs[lowerModuleIndex] >= N_MAX_MD_PER_MODULES)
                 {
-                    #ifdef Warnings
-                    if(mdModuleIndex == N_MAX_MD_PER_MODULES)
-                        printf("Mini-doublet excess alert! Module index =  %d\n",lowerModuleIndex);
-                    #endif
+#ifdef Warnings
+                    printf("Mini-doublet excess alert! Module index =  %d\n",lowerModuleIndex);
+#endif
                 }
                 else
                 {
+                    unsigned int mdModuleIndex = atomicAdd(&mdsInGPU.nMDs[lowerModuleIndex],1);
```

Comment:
I think that this part is incorrect: `mdModuleIndex` should be computed first so that it's a fixed value in a current thread. As written, between a load on L72 where the check is made and the `atomicAdd` call the value could've been already incremented in another thread.
So, an occasional over-write is possible, especially in region1 (2S barrel layers).
Also, it's even more likely that `totOccupancyMDs` and `nMDs` are out of sync in these cases.

Reply:
I'm not sure if this has already been said, but couldn't we put the check on totOccupancy instead of nMDs. As I understand, currently two threads can run the nMD check and then both would pass and increment the counter. So you need the atomic add to happen before the check. But then nMDs would keep incremented (Even when the value isnt saved). But just put the check on totOccupancy which you want to keep incrementing anyway. then nMDs inside the if statement would not be incremented if totOCcupancy is already above the Max and those values would not be written either.

Reply:
I don't think this has been already mentioned. I am not sure that it will work but I can try for sure. Thanks for the suggestion.

Reply:
It may work.

I could imagine the following though.
 Let's say before two concurrent threads A and B tot = M-3, nM=M-3 (where M is max).
- A: tot_new = M-1 (happened second), nM_new = M-2 (happened first)
- B: tot_new = M-2 (happened first), nM_new=M-1 (happened second)

but I guess we do not care much if tot and nM can be out of sync.